### PR TITLE
Sous Bootstrap command

### PR DESCRIPTION
**Pull request by nyarly**
You *should* be able to `go get github.com/opentable/sous/cmd/sous-bootstrap`
and then use that command to do bootstrapping when e.g. you bork the dev
servers.

Worst case, go to your Sous checkout, then:
```
cd cmd/sous-bootstrap
dep ensure
go install .
```

I'd be okay with that, if it's necessary, because it's not a general-purpose command.

### CODE REVIEW

**Comment by sous-machine-account** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR: 42342 lines of code changed</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: cmd/sous-bootstrap/bootstrap.go [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+			log.Printf("  %s up and running: %v", bcfg.localListen, c)
+			return nil
+		}
+		if _, ok := err.(*net.OpError); !ok {
+			log.Print(err)
+			return err
+		}
+		time.Sleep(250 * time.Millisecond)
+	}
+	err := fmt.Errorf("Server didn't start in 30 seconds.")
```
[the remark this answers is not in the document]
Comment (nyarly) [reply cmd/sous-bootstrap/bootstrap.go]: Used directly as output

File: cmd/sous-bootstrap/bootstrap.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	bootstrap.Flags().StringVarP(&bcfg.serverCfgPath, "server-config", "C", "", "the path to the ephemeral server's config directory")
+	bootstrap.Flags().StringVarP(&bcfg.repo, "repo", "r", "", "the repo of the bootstrapped service")
+	bootstrap.Flags().StringVarP(&bcfg.offset, "offset", "o", "", "the offset of the bootstrapped service")
+	bootstrap.Flags().StringVarP(&bcfg.flavor, "flavor", "f", "", "the flavor of the bootstrapped service")
+	bootstrap.Flags().StringVarP(&bcfg.tag, "tag", "t", "", "the tag of the bootstrapped service")
+	bootstrap.Flags().StringVarP(&bcfg.localListen, "listen", "l", "localhost:61000", "the address for the ephemeral to listen on")
+}
+
+func verifyConfig(*cobra.Command, []string) {
+	for _, pair := range []struct{ f, v string }{
```
Comment (lachlanh) [inline cmd/sous-bootstrap/bootstrap.go]: hard to read

File: cmd/sous-bootstrap/bootstrap.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+	}
+}
+
+func runBootstrap() error {
+	server, err := runServer()
+	if err != nil {
+		return err
+	}
+	log.Printf("Server pid: %d", server.Process.Pid)
+	defer func(server *exec.Cmd) {
```
Comment (lachlanh) [inline cmd/sous-bootstrap/bootstrap.go]: split out ?

File: cmd/sous-bootstrap/bootstrap.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+
+func runDeploy() error {
+	extraArgs := []string{}
+	if bcfg.offset != "" {
+		extraArgs = append(extraArgs, "-offset="+bcfg.offset)
+	}
+	if bcfg.flavor != "" {
+		extraArgs = append(extraArgs, "-flavor="+bcfg.flavor)
+	}
+	deploy := exec.Command("sous", append([]string{"deploy", "-cluster=" + bcfg.cluster, "-repo=" + bcfg.repo, "-tag=" + bcfg.tag}, extraArgs...)...)
```
Comment (lachlanh) [inline cmd/sous-bootstrap/bootstrap.go]: hard to read.

**Comment by lachlanh** [comment]
Looks good, hard to read comments are to highlight where I had to pause and think about the formulation and what was going on. no need to change.

File: cmd/sous-bootstrap/bootstrap.go (hunk 4 above)
Comment (lachlanh) [reply cmd/sous-bootstrap/bootstrap.go]: ```args := append([]string{"deploy", "-cluster=" + bcfg.cluster, "-repo=" + bcfg.repo, "-tag=" + bcfg.tag}, extraArgs...)```
```deploy := exec.Command("sous", args...)```

File: cmd/sous-bootstrap/bootstrap.go (hunk 2 above)
Comment (lachlanh) [reply cmd/sous-bootstrap/bootstrap.go]: could split the struct definition out, or may be a formulation without the loop ?
{"cluster", exitOnEmpty(bcfg.cluster)}

**Pull request merged**